### PR TITLE
[chore][exporter/azuremonitor] Force the telemetry to be sent to azuremonitor

### LINKED ISSUE #35037: Istio Traces Aren't Being Exported with Azure Monitor Exporter
### Component(s)

exporter/azuremonitor

## Description

When using Istio with the OpenTelemetry Azure Monitor exporter, traces are not being propagated to Azure Monitor. 

## Steps to Reproduce

1. Set up an Istio service mesh with tracing enabled.
2. Deploy the OpenTelemetry Collector with the azuremonitor exporter to send trace data to Application Insights.
3. Generate traffic within the mesh and observe the trace data in Application Insights

## Expected Result
Traces generated by Istio should appear correctly in Application Insights with proper formatting and trace context propagation using W3C Trace Context headers.

## Actual Result
Traces are not appearing in Application Insights suggesting a failure to convert B3 headers into the W3C Trace Context format. No logging provided that shows the trace data was rejected by Application Insights

### Collector version

923eb1cf

## Environment
OS: Ubuntu 22.04.04

## Istio installed with Helm 

helm repo add istio https://istio-release.storage.googleapis.com/charts
helm repo update

helm install istio-base istio/base -n istio-system --create-namespace
helm install istiod istio/istiod -n istio-system --wait
helm status istiod -n istio-system

### Configure Providers 

kubectl get configmap istio -n istio-system -o yaml > configmap.yaml

### Update configmap for grpc OTLP format Traces
  
  mesh: |-
    defaultConfig:
      discoveryAddress: istiod.istio-system.svc:15012
      tracing: {}
    defaultProviders:
      metrics:
        - prometheus
    enablePrometheusMerge: true
    rootNamespace: istio-system
    trustDomain: cluster.local
    enableTracing: true
    extensionProviders:
      - name: otel-tracing
        opentelemetry:
          port: 4317
          service: opentelemetry-collector.otel.svc.cluster.local
          grpc: {} 

kubectl apply -f configmap.yaml

## Install Otel Collector
kubectl create namespace otel
kubectl label namespace otel istio-injection=enabled

```
cat <<EOF > otel-collector-contrib.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: opentelemetry-collector-conf
  labels:
    app: opentelemetry-collector
data:
  opentelemetry-collector-config: |
    receivers:
      otlp:
        protocols:
          grpc:
          http:
    processors:
      batch:
    exporters:
      logging:
        loglevel: debug
      azuremonitor:
        connection_string: "InstrumentationKey="
        spaneventsenabled: true
        maxbatchinterval: .05s
        sending_queue:
          enabled: true
          num_consumers: 10
          queue_size: 2
    extensions:
      health_check:
        port: 13133
    service:
      extensions:
      - health_check
      telemetry:
        logs:
          debug:
            verbosity: detailed
      pipelines:
        logs:
          receivers: [otlp]
          processors: [batch]
          exporters: [logging, azuremonitor]
        traces:
          receivers: [otlp]
          exporters: [logging, azuremonitor]
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: opentelemetry-collector
spec:
  selector:
    matchLabels:
      app: opentelemetry-collector
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: opentelemetry-collector
    spec:
      containers:
        - name: opentelemetry-collector
          image: otel/opentelemetry-collector-contrib:latest
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 4317
              protocol: TCP
            - containerPort: 4318
              protocol: TCP
          resources:
            limits:
              cpu: "2"
              memory: 4Gi
            requests:
              cpu: 200m
              memory: 400Mi
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - name: opentelemetry-collector-config-vol
              mountPath: /etc/otel
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      terminationGracePeriodSeconds: 30
      volumes:
        - configMap:
            defaultMode: 420
            items:
              - key: opentelemetry-collector-config
                path: opentelemetry-collector-config.yaml
            name: opentelemetry-collector-conf
          name: opentelemetry-collector-config-vol
---
apiVersion: v1
kind: Service
metadata:
  name: opentelemetry-collector
  labels:
    app: opentelemetry-collector
spec:
  ports:
    - name: grpc-otlp # Default endpoint for OpenTelemetry receiver.
      port: 4317
      protocol: TCP
      targetPort: 4317
    - name: http-otlp # HTTP endpoint for OpenTelemetry receiver.
      port: 4318
      protocol: TCP
      targetPort: 4318
  selector:
    app: opentelemetry-collector
EOF
```

kubectl apply -f otel-collector-contrib.yaml -n otel

## Set up demo

kubectl create namespace demo
kubectl label namespace demo istio-injection=enabled

### Create Telemetry Rule

cat <<EOF > tel-rule-otel-tracing.yaml
apiVersion: telemetry.istio.io/v1
kind: Telemetry
metadata:
  name: otel-tracing
  namespace: demo
spec:
  tracing:
  - providers:
    - name: otel-tracing
    randomSamplingPercentage: 100
    customTags:
      "app-insights":
        literal:
          value: "from-otel-collector"
EOF

kubectl apply -f tel-rule-otel-tracing.yaml

### Generate and view traces 
kubectl apply -f https://raw.githubusercontent.com/istio/istio/master/samples/bookinfo/platform/kube/bookinfo.yaml -n demo
kubectl get pods -n demo

## Generate traces 
for i in $(seq 1 100); do kubectl exec "$(kubectl get pod -l app=ratings -o jsonpath='{.items[0].metadata.name}' -n demo)" -c ratings -n demo -- curl -sS productpage:9080/productpage | grep -o "<title>.*</title>"; done

## Verify traces on console 
  
kubectl logs -n otel "$(kubectl get pods -n otel -l app=opentelemetry-collector -o jsonpath='{.items[0].metadata.name}')" | grep "app-insights"

### OpenTelemetry Collector configuration

```yaml
receivers:
      otlp:
        protocols:
          grpc:
          http:
    processors:
      batch:
    exporters:
      logging:
        loglevel: debug
      azuremonitor:
        connection_string: "InstrumentationKey="
        spaneventsenabled: true
        maxbatchinterval: .05s
        sending_queue:
          enabled: true
          num_consumers: 10
          queue_size: 2
    extensions:
      health_check:
        port: 13133
    service:
      extensions:
      - health_check
      telemetry:
        logs:
          debug:
            verbosity: detailed
      pipelines:
        logs:
          receivers: [otlp]
          processors: [batch]
          exporters: [logging, azuremonitor]
        traces:
          receivers: [otlp]
          exporters: [logging, azuremonitor]
```

### Log output

```shell
2024-09-04T15:46:16.212Z    info    TracesExporter    {"kind": "exporter", "data_type": "traces", "name": "debug", "resource spans": 1, "spans": 1}
2024-09-04T15:46:16.212Z    info    ResourceSpans #0
Resource SchemaURL: 
Resource attributes:
     -> service.name: Str(details.demo)
ScopeSpans #0
ScopeSpans SchemaURL: 
InstrumentationScope  
Span #0
    Trace ID       : e7e12ca17cf7655ff46ecd79cec9451c
    Parent ID      : 9e8b852dba31ab84
    ID             : 05226d559db0c33d
    Name           : details.demo.svc.cluster.local:9080/*
    Kind           : Server
    Start time     : 2024-09-04 15:46:14.652164 +0000 UTC
    End time       : 2024-09-04 15:46:14.655316 +0000 UTC
    Status code    : Unset
    Status message : 
Attributes:
     -> node_id: Str(sidecar~10.244.0.13~details-v1-79dfbd6fff-j7jjs.demo~demo.svc.cluster.local)
     -> zone: Str()
     -> guid:x-request-id: Str(d17c4164-5765-9003-a163-04039b0b98e0)
     -> http.url: Str(http://details:9080/details/0)
     -> http.method: Str(GET)
     -> downstream_cluster: Str(-)
     -> user_agent: Str(curl/7.88.1)
     -> http.protocol: Str(HTTP/1.1)
     -> peer.address: Str(10.244.0.18)
     -> request_size: Str(0)
     -> response_size: Str(178)
     -> component: Str(proxy)
     -> upstream_cluster: Str(inbound|9080||)
     -> upstream_cluster.name: Str(inbound|9080||;)
     -> http.status_code: Str(200)
     -> response_flags: Str(-)
     -> istio.mesh_id: Str(cluster.local)
     -> istio.canonical_revision: Str(v1)
     -> istio.canonical_service: Str(details)
     -> app-insights: Str(otel)
     -> istio.cluster_id: Str(Kubernetes)
     -> istio.namespace: Str(demo)
    {"kind": "exporter", "data_type": "traces", "name": "debug"}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by whitneygriffith**
#### Description

1. Resolved an issue where traces weren't being sent to App Insights due to not flushing the Telemetry Channel. Added the necessary flush operation to ensure all traces, metrics and logs are properly sent to the queue, leveraging App Insights' batch handling for more efficient processing.

#### Link to tracking issue
Resolves #35037

### CODE REVIEW

**Review by pcwiese — APPROVED** [review]

**Comment by andrzej-stencel** [comment]
Sounds like this is a bug fix. Can you add a changelog entry? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

**Comment by whitneygriffith** [comment]
> Sounds like this is a bug fix. Can you add a changelog entry? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

I have added the changelog entry.

File: .chloggen/azuremonitor-bug-fix.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'bug_fix'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: azuremonitor
```
Comment (andrzej-stencel) [inline .chloggen/azuremonitor-bug-fix.yaml:7]:
```suggestion
component: azuremonitorexporter
```

**Review by andrzej-stencel — APPROVED** [review]

**Comment by andrzej-stencel** [comment]
There are some errors here: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11700948723/job/32586089416?pr=36172

**Review by fatsheep9146 — COMMENTED** [review]
There are still some unresolved failed tests https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11711127058/job/32641600088?pr=36172
```
=== FAIL: . TestExporterLogDataCallback (0.00s)
panic: 
	assert: mock: I don't know what to return because the method call was unexpected.
		Either do Mock.On("Flush").Return(...) first, or remove the Flush() call.
		This method was unexpected:
			Flush()
			
		at: [/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/mock_transportChannel.go:24 /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/logexporter.go:39 /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/logexporter_test.go:118] [recovered]
	panic: 
	assert: mock: I don't know what to return because the method call was unexpected.
		Either do Mock.On("Flush").Return(...) first, or remove the Flush() call.
		This method was unexpected:
			Flush()
			
		at: [/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/mock_transportChannel.go:24 /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/logexporter.go:39 /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/azuremonitorexporter/logexporter_test.go:118]
```

**Review by fatsheep9146 — COMMENTED** [review]
I think you should do as this lint said, could you please fix this? https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11904647742/job/33173837807?pr=36172
```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```

**Review by fatsheep9146 — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Comment by codeboten** [comment]
Closing and re-opening as it looks like the CI checks are stuck.

**Comment by codeboten** [comment]
Apparently I cannot reopen it :(

Screenshot 2024-11-22 at 2 52 22 PM

**Comment by fatsheep9146** [comment]
@whitneygriffith How about you create a same pr again? It seems that this pr is not able to be merged due to the stuck CI.

**Comment by whitneygriffith** [comment]
Created #36520

**Pull request closed without merging**